File: leetcode_C++/p2596.cpp
/*
@Author: mrxhar
@Date:   2023-09-13 08:41:50
@Email:  1337904145@qq.com
@Blog:   https://leetcode.cn/problems/check-knight-tour-configuration/?envType=daily-question&envId=2023-09-13
*/
#include <unordered_set>
#include <vector>
using std::vector;
using namespace std;

class Solution {
public:
    bool checkValidGrid(vector<vector<int>>& grid) {
        if (grid[0][0] != 0)
        {
            return false;
        }
        
        int n = grid.size();
        vector<pair<int, int>> p(n * n, pair<int, int>());
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                p[grid[i][j]] = {i, j};
            }
        }
        for (int i = 1; i < n * n; i++) {
            int x = abs(p[i].first - p[i - 1].first), y = abs(p[i].second - p[i - 1].second);
            if (x * y != 2) {
                return false;
            }
        }
        return true;
    }
};

File: leetcode_py/p2596.py
from typing import List


class Solution:
    def checkValidGrid(self, grid: List[List[int]]) -> bool:
        if grid[0][0]:
            return False

        n = len(grid)
        index = [[] for _ in range(n * n)]
        for i in range(n):
            for j in range(n):
                index[grid[i][j]] = [i, j]
        for i in range(1, n * n):
            x = abs(index[i][0] - index[i - 1][0])
            y = abs(index[i][1] - index[i - 1][1])
            if x * y != 2:
                return False
        return True
